## 1. Simple Linear Regression With scikit-learn
First, we need to import the appropriate packagaes and classes, which in this case are numpy and LinearRegression from sklearn.linear_model. 

In [1]:
import numpy as np
from sklearn.linear_model import LinearRegression

Second we need to define the data we will be working with; the inputs or regressors correspond to x and the outputs or predictors correspond to y. Both of them should be arrays or similar objects. 

In [2]:
x = np.array([5, 15, 25, 35, 45, 55]).reshape((-1, 1))
y = np.array([5, 20, 14, 32, 22, 38])

We use the reshape() fucntion on x because this array needs to be two-dimensional, i.e. to have one column and as many rows as necessary. That is what the argument (-1, 1) of the function specifies. Then we need to take a look at the data, so we print it. 

In [4]:
print(x)
print(y)

[[ 5]
 [15]
 [25]
 [35]
 [45]
 [55]]
[ 5 20 14 32 22 38]


After that we create a model and fit it with LinearRegression(). The default arguments are: 
- fit_intercept is a Boolean (True by default) that decides whether to calculate the intercept 𝑏₀ (True) or consider it equal to zero (False).
- normalize is a Boolean (False by default) that decides whether to normalize the input variables (True) or not (False).
- copy_X is a Boolean (True by default) that decides whether to copy (True) or overwrite the input variables (False).
- n_jobs is an integer or None (default) and represents the number of jobs used in parallel computation. None usually means one job and -1 to use all processors.

In [5]:
model = LinearRegression()

Then, with .fit(), we calculate the optimal values of the weights 𝑏₀ and 𝑏₁, using the existing input and output (x and y) as the arguments, i.e. we fit the model. We can do both of this steps in one with the following code:

In [6]:
model = LinearRegression().fit(x, y)

After fitting the model, we can get the results to make sure it is functional. In this case, the arguments are x and y. 

In [7]:
r_sq = model.score(x, y)
print('coefficient of determination:', r_sq)

coefficient of determination: 0.715875613747954


The attributes of the model are .intercept_, which represents the coefficient, 𝑏₀ (y intercept) and .coef_, which represents 𝑏1 (the slope):

In [9]:
print('intercept:', model.intercept_)
print('slope:', model.coef_)

intercept: 5.633333333333329
slope: [0.54]


Take notice that .intercept_ is a scalar, while .coef_ is an array. This results tell us that the model predicts the response 5.63 when 𝑥 is zero. The value 𝑏₁ = 0.54 means that the predicted response rises by 0.54 when 𝑥 is increased by one. We can provide y as a 2-dimensional array by doing the following, which returns similar results:

In [10]:
new_model = LinearRegression().fit(x, y.reshape((-1, 1)))
print('intercept:', new_model.intercept_)
print('slope:', new_model.coef_)

intercept: [5.63333333]
slope: [[0.54]]


To predict responses we use the predict() function, which passes the regressor as the argument and returns the corresponding predicted response: 

In [12]:
y_pred = model.predict(x)
print('predicted response:', y_pred, sep='\n')

predicted response:
[ 8.33333333 13.73333333 19.13333333 24.53333333 29.93333333 35.33333333]


We can also do the following, which multiplies each element of x with model.coef_ and add model.intercept_ to the product. The predicted response is now a two-dimensional array:

In [13]:
y_pred = model.intercept_ + model.coef_ * x
print('predicted response:', y_pred, sep='\n')

predicted response:
[[ 8.33333333]
 [13.73333333]
 [19.13333333]
 [24.53333333]
 [29.93333333]
 [35.33333333]]


Since regression models are often applied for forecasts, we can use fitted models to calculate the outputs based on some other, new inputs. In the following example, .predict() is applied to the new regressor x_new and yields the response y_new. This example conveniently uses arange() from numpy to generate an array with the elements from 0 (inclusive) to 5 (exclusive), i.e. 0,1,2,3,4. 

In [14]:
x_new = np.arange(5).reshape((-1, 1))
print(x_new)

[[0]
 [1]
 [2]
 [3]
 [4]]


In [15]:
y_new = model.predict(x_new)
print(y_new)

[5.63333333 6.17333333 6.71333333 7.25333333 7.79333333]


## 2. Multiple Linear Regression With scikit-learn
As before, we first import numpy and LinearRegression from sklearn.linear_model. We will not doing again as these packages have been imported already. We then provide known inputs and outputs and print them to make sure they are what we want. In this type of regression, x is a two-dimensional array with at least two columns (exactly two in this case), while y is usually a one-dimensional array.

In [17]:
x = [[0, 1], [5, 1], [15, 2], [25, 5], [35, 11], [45, 15], [55, 34], [60, 35]]
y = [4, 5, 20, 14, 32, 22, 38, 43]
x, y = np.array(x), np.array(y)
print(x)
print(y)

[[ 0  1]
 [ 5  1]
 [15  2]
 [25  5]
 [35 11]
 [45 15]
 [55 34]
 [60 35]]
[ 4  5 20 14 32 22 38 43]


Then, we fit the model and get its properties as we did before:

In [18]:
model = LinearRegression().fit(x, y)

In [19]:
r_sq = model.score(x, y)
print('coefficient of determination:', r_sq)
print('intercept:', model.intercept_)
print('slope:', model.coef_)

coefficient of determination: 0.8615939258756776
intercept: 5.52257927519819
slope: [0.44706965 0.25502548]


We get 𝑅² using .score() and the values of the estimators of regression coefficients with .intercept_ (𝑏₀) and .coef_ (𝑏₁ and 𝑏₂). In this example, the intercept is approximately 5.52, i.e. the value of the predicted response when 𝑥₁ = 𝑥₂ = 0. The increase of 𝑥₁ by 1 yields the rise of the predicted response by 0.45. Similarly, when 𝑥₂ grows by 1, the response rises by 0.26.

Predictions work the same way as with simple linear regression: 

In [20]:
y_pred = model.predict(x)
print('predicted response:', y_pred, sep='\n')

predicted response:
[ 5.77760476  8.012953   12.73867497 17.9744479  23.97529728 29.4660957
 38.78227633 41.27265006]


We can also do it the following way, which yields similar results. In this case, we predict the output values by multiplying each column of the input with the appropriate weight, summing the results and adding the intercept to the sum.

In [21]:
y_pred = model.intercept_ + np.sum(model.coef_ * x, axis=1)
print('predicted response:', y_pred, sep='\n')

predicted response:
[ 5.77760476  8.012953   12.73867497 17.9744479  23.97529728 29.4660957
 38.78227633 41.27265006]


Finally, we can apply this model to new data as well:

In [22]:
x_new = np.arange(10).reshape((-1, 2))
print(x_new)

[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


In [23]:
y_new = model.predict(x_new)
print(y_new)

[ 5.77760476  7.18179502  8.58598528  9.99017554 11.3943658 ]


## 3. Polynomial Regression With scikit-learn
In this case, we need the same packages as before plus PolynomialFeatures class from sklearn.preprocessing: 

In [24]:
from sklearn.preprocessing import PolynomialFeatures

We define the regressors and predictors are defined the same way as before. Since we need the input to be a two-dimensional array, we use .reshape(). 

In [25]:
x = np.array([5, 15, 25, 35, 45, 55]).reshape((-1, 1))
y = np.array([15, 11, 2, 8, 25, 32])

There is a new step when doing polynomial regression. We need to include 𝑥² as an additional feature, so we should transform the input array x to contain the additional column(s) with the values of 𝑥² (and eventually other features if needed). The class PolynomialFeatures is very convenient for this purpose. In this case the variable transformer refers to an instance of PolynomialFeatures which you can use to transform the input x. The defaults are:
- degree is an integer (2 by default) that represents the degree of the polynomial regression function.
- interaction_only is a Boolean (False by default) that decides whether to include only interaction features (True) or all features (False).
- include_bias is a Boolean (True by default) that decides whether to include the bias (intercept) column of ones (True) or not (False).

In [26]:
transformer = PolynomialFeatures(degree=2, include_bias=False)

Once we have the transformer variable set, we need to fit it: 

In [27]:
transformer.fit(x)

PolynomialFeatures(degree=2, include_bias=False, interaction_only=False,
                   order='C')

The, transformer is ready to create a new, modified input, applying .transform(), which takes the input array as the argument and returns the modified array.

In [29]:
x_ = transformer.transform(x)

To do the three past statements in one, we can do the following: 

In [38]:
x_ = PolynomialFeatures(degree=2, include_bias=False).fit_transform(x)

The modified array will look the following way, one column with the original inputs and the other with their squares.

In [39]:
print(x_)

[[   5.   25.]
 [  15.  225.]
 [  25.  625.]
 [  35. 1225.]
 [  45. 2025.]
 [  55. 3025.]]


We then fit the model as we did before:

In [40]:
model = LinearRegression().fit(x_, y)

We obtain the properties of the model, the same way as we did before:

In [41]:
r_sq = model.score(x_, y)
print('coefficient of determination:', r_sq)
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

coefficient of determination: 0.8908516262498564
intercept: 21.372321428571425
coefficients: [-1.32357143  0.02839286]


As before: .score() returns 𝑅². Its first argument is also the modified input x_, not x. The values of the weights are associated to .intercept_ and .coef_: .intercept_ represents 𝑏₀, while .coef_ references the array that contains 𝑏₁ and 𝑏₂ respectively.

We can obtain similar results, with different transformation and regression arguments, like setting the include_bias to true, which yields three columns: the first column of x_ contains ones, the second has the values of x, while the third holds the squares of x.

In [43]:
x_ = PolynomialFeatures(degree=2, include_bias=True).fit_transform(x)
print(x_)

[[1.000e+00 5.000e+00 2.500e+01]
 [1.000e+00 1.500e+01 2.250e+02]
 [1.000e+00 2.500e+01 6.250e+02]
 [1.000e+00 3.500e+01 1.225e+03]
 [1.000e+00 4.500e+01 2.025e+03]
 [1.000e+00 5.500e+01 3.025e+03]]


Since the intercept is already included with the leftmost column of ones, we don’t need to include it again when creating the instance of LinearRegression, meaning we can provide fit_intercept=False, the following way:

In [44]:
model = LinearRegression(fit_intercept=False).fit(x_, y)

We then get the results, which are similar to the ones we had before. Now .intercept_ is zero, but .coef_ actually contains 𝑏₀ as its first element: 

In [45]:
r_sq = model.score(x_, y)
print('coefficient of determination:', r_sq)
print('intercept:', model.intercept_)
print('coefficients:', model.coef_)

coefficient of determination: 0.8908516262498565
intercept: 0.0
coefficients: [21.37232143 -1.32357143  0.02839286]


To get a predicted response, we do the same as before, but must include the x_ argument instead:

In [46]:
y_pred = model.predict(x_)
print('predicted response:', y_pred, sep='\n')

predicted response:
[15.46428571  7.90714286  6.02857143  9.82857143 19.30714286 34.46428571]


We can follow a similar procedure with several input variables. We'll have an input array with more than one column, but everything else is the same:

In [47]:
# Step 1: Import packages
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

# Step 2a: Provide data
x = [[0, 1], [5, 1], [15, 2], [25, 5], [35, 11], [45, 15], [55, 34], [60, 35]]
y = [4, 5, 20, 14, 32, 22, 38, 43]
x, y = np.array(x), np.array(y)

# Step 2b: Transform input data
x_ = PolynomialFeatures(degree=2, include_bias=False).fit_transform(x)

# Step 3: Create a model and fit it
model = LinearRegression().fit(x_, y)

# Step 4: Get results
r_sq = model.score(x_, y)
intercept, coefficients = model.intercept_, model.coef_

# Step 5: Predict
y_pred = model.predict(x_)

This regression example yields the following results and predictions:

In [49]:
print('coefficient of determination:', r_sq)
print('intercept:', intercept)
print('coefficients:', coefficients, sep='\n')
print('predicted response:', y_pred, sep='\n')

coefficient of determination: 0.9453701449127822
intercept: 0.8430556452395734
coefficients:
[ 2.44828275  0.16160353 -0.15259677  0.47928683 -0.4641851 ]
predicted response:
[ 0.54047408 11.36340283 16.07809622 15.79139    29.73858619 23.50834636
 39.05631386 41.92339046]


In this case, there are six regression coefficients (including the intercept), as shown in the estimated regression function 𝑓(𝑥₁, 𝑥₂) = 𝑏₀ + 𝑏₁𝑥₁ + 𝑏₂𝑥₂ + 𝑏₃𝑥₁² + 𝑏₄𝑥₁𝑥₂ + 𝑏₅𝑥₂².

You can also notice that polynomial regression yielded a higher coefficient of determination than multiple linear regression for the same problem. At first, you could think that obtaining such a large 𝑅² is an excellent result. It might be.

However, in real-world situations, having a complex model and 𝑅² very close to 1 might also be a sign of overfitting. To check the performance of a model, you should test it with new data, that is with observations not used to fit (train) the model.

## 4. Advanced Linear Regression With statsmodels
We can implement linear regression in Python relatively easily by using the package statsmodels as well. We first need to import it: 

In [50]:
import numpy as np
import statsmodels.api as sm

/Users/LEBC/anaconda3/lib/python3.7/site-packages/statsmodels/compat/pandas.py:23: FutureWarning: The Panel class is removed from pandas. Accessing it from the top-level namespace will also be removed in the next version
  data_klasses = (pandas.Series, pandas.DataFrame, pandas.Panel)


We then provide data and transform inputs: 

In [51]:
x = [[0, 1], [5, 1], [15, 2], [25, 5], [35, 11], [45, 15], [55, 34], [60, 35]]
y = [4, 5, 20, 14, 32, 22, 38, 43]
x, y = np.array(x), np.array(y)

The input and output arrays are created this way, but then we need to add the column of ones to the inputs if you want statsmodels to calculate the intercept 𝑏₀ since it doesn’t takes 𝑏₀ into account by default.

In [52]:
x = sm.add_constant(x)

You add the column of ones to x with add_constant(), which takes the input array x as an argument and returns a new array with the column of ones inserted at the beginning. This is how x and y look like now:

In [53]:
print(x)

[[ 1.  0.  1.]
 [ 1.  5.  1.]
 [ 1. 15.  2.]
 [ 1. 25.  5.]
 [ 1. 35. 11.]
 [ 1. 45. 15.]
 [ 1. 55. 34.]
 [ 1. 60. 35.]]


In [54]:
print(y)

[ 4  5 20 14 32 22 38 43]


The regression model based on ordinary least squares is an instance of the class statsmodels.regression.linear_model.OLS. The first argument is the output, followed with the input.

In [58]:
model = sm.OLS(y, x)

We fit the model as we did before: 

In [59]:
results = model.fit()

Finally, to get the results we just use the summary() function: 

In [57]:
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.862
Model:                            OLS   Adj. R-squared:                  0.806
Method:                 Least Squares   F-statistic:                     15.56
Date:                Sat, 05 Oct 2019   Prob (F-statistic):            0.00713
Time:                        16:05:26   Log-Likelihood:                -24.316
No. Observations:                   8   AIC:                             54.63
Df Residuals:                       5   BIC:                             54.87
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.5226      4.431      1.246      0.2

/Users/LEBC/anaconda3/lib/python3.7/site-packages/scipy/stats/stats.py:1450: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=8
  "anyway, n=%i" % int(n))


To extract values from the table above, we can do the following, where:
1. .rsquared holds 𝑅².
2. .rsquared_adj represents adjusted 𝑅² (𝑅² corrected according to the number of input features).
3. .params refers the array with 𝑏₀, 𝑏₁, and 𝑏₂ respectively.

In [61]:
print('coefficient of determination:', results.rsquared)
print('adjusted coefficient of determination:', results.rsquared_adj)
print('regression coefficients:', results.params)

coefficient of determination: 0.8615939258756777
adjusted coefficient of determination: 0.8062314962259488
regression coefficients: [5.52257928 0.44706965 0.25502548]


We can obtain the predicted response on the input values used for creating the model using .fittedvalues or .predict() with the input array as the argument:

In [62]:
# fittedvalues
print('predicted response:', results.fittedvalues, sep='\n')
# predict
print('predicted response:', results.predict(x), sep='\n')

predicted response:
[ 5.77760476  8.012953   12.73867497 17.9744479  23.97529728 29.4660957
 38.78227633 41.27265006]
predicted response:
[ 5.77760476  8.012953   12.73867497 17.9744479  23.97529728 29.4660957
 38.78227633 41.27265006]


If you want predictions with new regressors, you can also apply .predict() with new data as the argument:

In [63]:
x_new = sm.add_constant(np.arange(10).reshape((-1, 2)))
print(x_new)

[[1. 0. 1.]
 [1. 2. 3.]
 [1. 4. 5.]
 [1. 6. 7.]
 [1. 8. 9.]]


In [64]:
y_new = results.predict(x_new)
print(y_new)

[ 5.77760476  7.18179502  8.58598528  9.99017554 11.3943658 ]
